diesel/src/query_builder/insert_statement.rs

```diff
@@ -73,23 +79,26 @@ impl<T, U, Op, DB> QueryFragment<DB> for InsertStatement<T, U, Op> where
     }
 }
 
-impl_query_id!(noop: InsertStatement<T, U, Op>);
+impl_query_id!(noop: InsertStatement<T, U, Op, Ret>);
 
-impl<T, U, Op> AsQuery for InsertStatement<T, U, Op> where
+impl<T, U, Op> AsQuery for InsertStatement<T, U, Op, NoReturningClause> where
     T: Table,
-    InsertQuery<T::AllColumns, InsertStatement<T, U, Op>>: Query,
+    InsertStatement<T, U, Op, ReturningClause<T::AllColumns>>: Query,
 {
     type SqlType = <Self::Query as Query>::SqlType;
-    type Query = InsertQuery<T::AllColumns, Self>;
+    type Query = InsertStatement<T, U, Op, ReturningClause<T::AllColumns>>;
 
     fn as_query(self) -> Self::Query {
-        InsertQuery {
-            returning: T::all_columns(),
-            statement: self,
-        }
+        self.returning(T::all_columns())
     }
 }
 
+impl<T, U, Op, Ret> Query for InsertStatement<T, U, Op, ReturningClause<Ret>> where
+    Ret: Expression + SelectableExpression<T> + NonAggregate,
+{
+    type SqlType = Ret::SqlType;
+}
+
 impl<T, U, Op> InsertStatement<T, U, Op> {
```

Comment:
This is the magic right here, isn't it? As this assumes the default type param for `Ret` it is only implemented for `NoReturningClause`, right? And thus not implemented once you called `returning`. Do you want to explicitly write `NoReturningClause` here and maybe add a short doc comment before the impl to explain this?